ci_scripts/utils/configuration.py

```diff
@@ -0,0 +1,276 @@
+        self._raise_undefined(key)
+
+    def _raise_undefined(self, key: Optional[str]):
+        raise Undefined(f'Undefined key: {key}')
+
+    def get_value(self, key: Union[str, ConfigurationVariable]) -> Any:
+        """Gets a configuration value.
+
+        If the variable was not defined, an exception is raised.
+
+        Args:
+            key: variable key. This can be a string or a ConfigurationVariable
+            element.
+
+        Returns:
+            configuration value corresponding to the key.
+        """
+        if not key:
+            raise KeyError(key)
+        key_str = key.name if isinstance(key, ConfigurationVariable) else key
+        return self._fetch_value(key_str)
+
+    def get_value_or_default(self, key: Union[str, ConfigurationVariable],
+                             default_value: Any) -> Any:
+        """Gets a configuration value.
+
+        If the variable was not defined, the default value is returned.
+
+        Args:
+            key: variable key. This can be a string or a ConfigurationVariable
+            element.
+            default_value: value to default to if the variable was not defined.
+
+        Returns:
+            configuration value corresponding to the key.
+            default value if the variable is not defined.
+        """
+        try:
+            return self.get_value(key)
+        except Undefined as e:
+            logger.warning(e)
+            return default_value
+
+
+class StaticConfig(GenericConfig):
+    """Configuration with default values.
+
+    Only variables which are not likely do be different from a project to
+    another are defined here. They can be overridden by values in the
+    configuration file though. This should simply the number of variables
+    defined in toml.
+    """
+
+    BETA_BRANCH = 'beta'
+    MASTER_BRANCH = 'master'
+    RELEASE_BRANCH_PATTERN = r"^release.*$"
+    REMOTE_ALIAS = 'origin'
+    LOGGER_FORMAT = '%(levelname)s: %(message)s'
+    BOT_USERNAME = 'Monty Bot'
+    BOT_EMAIL = 'monty-bot@arm.com'
+    ORGANISATION = 'Arm Mbed'
+    ORGANISATION_EMAIL = 'support@mbed.com'
+
+    def _fetch_value(self, key: str) -> Any:
+        try:
+            return getattr(self, key)
+        except AttributeError:
+            self._raise_undefined(key)
+
+
+class EnvironmentConfig(GenericConfig):
+    """Configuration set in environment variables.
+
+    This also uses dotEnv mechanism.
+    """
+
+    def __init__(self):
+        """Constructor."""
+        dotenv.load_dotenv(
+            dotenv.find_dotenv(usecwd=True, raise_error_if_not_found=False))
+
+    def _fetch_value(self, key: str) -> Any:
+        environment_value = os.getenv(key)
+        if not environment_value:
+            self._raise_undefined(key)
+        return environment_value
+
+
+class FileConfig(GenericConfig):
+    """Configuration set in toml file."""
+
+    CONFIG_SECTION = 'ProjectConfig'
+    PATH_TOKEN = {'DIR', 'ROOT', 'PATH'}
```

Comment:
This seems like special magic that isn't documented when adding configuration variables.

Reply:
I wrote some more documentation about this behaviour